File: integrations/python/zep_adk/src/zep_adk/graph_search_tool.py
"""
ZepGraphSearchTool -- a model-callable ADK tool for searching Zep knowledge graphs.

Unlike ``ZepContextTool`` (which injects context automatically on every turn),
this tool is visible to the model in the tool list and called on-demand when
the model decides it needs to search the knowledge graph.

Search parameters can be *pinned* at construction time -- pinned parameters are
removed from the schema the model sees, locking them to a fixed value.  Any
parameter not pinned is exposed to the model with a reasonable default.

The tool resolves the search target automatically:

* If ``graph_id`` is set at construction → searches that shared graph for all
  users (e.g. a documentation knowledge base).
* If ``graph_id`` is not set → resolves ``user_id`` from ADK session state at
  runtime and searches the current user's personal graph.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from google.adk.tools.base_tool import BaseTool
from google.adk.tools.tool_context import ToolContext
from google.genai import types
from typing_extensions import override
from zep_cloud.client import AsyncZep

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Parameter definitions
# ---------------------------------------------------------------------------
# Each entry describes a graph.search parameter that can be pinned or exposed
# to the model.  Keys match the Zep SDK's ``graph.search()`` kwargs.

_SEARCH_PARAMS: dict[str, dict[str, Any]] = {
    "query": {
        "type": "STRING",
        "description": "Search query text (max 400 characters).",
        "required": True,
    },
    "scope": {
        "type": "STRING",
        "description": (
            "What to search for: 'edges' for facts and relationships, "
            "'nodes' for entity summaries, 'episodes' for raw messages."
        ),
        "enum": ["edges", "nodes", "episodes"],
        "default": "edges",
    },
    "reranker": {
        "type": "STRING",
        "description": (
            "Result ordering algorithm: 'rrf' (balanced), 'mmr' (diverse), "
            "'cross_encoder' (highest accuracy), 'episode_mentions' "
            "(frequently referenced), 'node_distance' (near a specific entity)."
        ),
        "enum": ["rrf", "mmr", "node_distance", "episode_mentions", "cross_encoder"],
        "default": "rrf",
    },
    "limit": {
        "type": "INTEGER",
        "description": "Maximum number of results to return.",
        "default": 10,
    },
    "mmr_lambda": {
        "type": "NUMBER",
        "description": (
            "Balance between diversity (0.0) and relevance (1.0). Only used when reranker is 'mmr'."
        ),
    },
    "center_node_uuid": {
        "type": "STRING",
        "description": (
            "UUID of the center node for distance-based reranking. "
            "Required when reranker is 'node_distance'."
        ),
    },
}

_TYPE_MAP: dict[str, types.Type] = {
    "STRING": types.Type.STRING,
    "INTEGER": types.Type.INTEGER,
    "NUMBER": types.Type.NUMBER,
    "BOOLEAN": types.Type.BOOLEAN,
}

# Parameters that are always constructor-only (complex types not suitable for
# model schema generation).
_CONSTRUCTOR_ONLY_PARAMS = frozenset({"search_filters", "bfs_origin_node_uuids"})

# All parameters that may be pinned at construction.
_PINNABLE_PARAMS = frozenset(_SEARCH_PARAMS.keys()) | _CONSTRUCTOR_ONLY_PARAMS


class ZepGraphSearchTool(BaseTool):
    """Model-callable tool for searching Zep knowledge graphs.

    This tool is added to the model's tool list and called on-demand when the
    model decides it needs to search for facts, entities, or prior messages.

    Any search parameter can be *pinned* at construction time by passing it as
    a keyword argument.  Pinned parameters are hidden from the model and locked
    to the given value.  Parameters not pinned are exposed in the model's tool
    schema with sensible defaults.

    Args:
        zep_client: An initialised ``AsyncZep`` client.
        graph_id: Optional fixed graph ID for shared-graph search.  When set,
            all searches target this graph regardless of which user's session
            is active.  When ``None``, the tool searches the current user's
            personal graph (resolved from session state).
        name: Tool name visible to the model.
        description: Tool description visible to the model.
        search_filters: Optional Zep search filters (constructor-only).
            Supports ``node_labels``, ``edge_types``, ``exclude_node_labels``,
            ``exclude_edge_types``, and property filters.
        bfs_origin_node_uuids: Optional list of node UUIDs for BFS seeding
            (constructor-only).
        **pinned: Any graph.search parameter to fix at construction time.
            Supported: ``scope``, ``reranker``, ``limit``, ``mmr_lambda``,
            ``center_node_uuid``.
    """

    def __init__(
        self,
        *,
        zep_client: AsyncZep,
        graph_id: str | None = None,
        name: str = "zep_graph_search",
        description: str = (
            "Search the user's knowledge graph for information from previous "
            "conversations, known facts about the user, or general context. "
            "Use this to look up specific details the user has shared before."
        ),
        search_filters: dict[str, Any] | None = None,
        bfs_origin_node_uuids: list[str] | None = None,
        **pinned: Any,
    ) -> None:
        super().__init__(name=name, description=description)
        self._zep: AsyncZep = zep_client
        self._graph_id: str | None = graph_id

        # Validate pinned params
        if "user_id" in pinned:
            raise ValueError(
                "'user_id' cannot be pinned. Per-user graph search is resolved "
                "from session state at runtime. Use 'graph_id' for shared "
                "graph search."
            )
        allowed_pinned = frozenset(_SEARCH_PARAMS.keys())
        unknown = set(pinned.keys()) - allowed_pinned
        if unknown:
            raise ValueError(
                f"Unknown pinned parameters: {unknown}. Allowed: {sorted(allowed_pinned)}"
            )

        # Store pinned search params
        self._pinned: dict[str, Any] = dict(pinned)
        if search_filters is not None:
            self._pinned["search_filters"] = search_filters
        if bfs_origin_node_uuids is not None:
            self._pinned["bfs_origin_node_uuids"] = bfs_origin_node_uuids

        # Pre-build the declaration once (it's immutable after construction)
        self._declaration = self._build_declaration()

    # ------------------------------------------------------------------
    # Schema declaration
    # ------------------------------------------------------------------

    def _build_declaration(self) -> types.FunctionDeclaration:
        """Build the function declaration, excluding pinned parameters."""
        properties: dict[str, types.Schema] = {}
        required: list[str] = []

        for param_name, param_def in _SEARCH_PARAMS.items():
            if param_name in self._pinned:
                continue  # pinned → hidden from model

            schema_kwargs: dict[str, Any] = {
                "type": _TYPE_MAP[param_def["type"]],
                "description": param_def.get("description", ""),
            }
            if "enum" in param_def:
                schema_kwargs["enum"] = param_def["enum"]

            properties[param_name] = types.Schema(**schema_kwargs)

            if param_def.get("required"):
                required.append(param_name)

        return types.FunctionDeclaration(
            name=self.name,
            description=self.description,
            parameters=types.Schema(
                type=types.Type.OBJECT,
                properties=properties,
                required=required or None,
            ),
        )

    @override
    def _get_declaration(self) -> types.FunctionDeclaration | None:
        return self._declaration

    # ------------------------------------------------------------------
    # Execution
    # ------------------------------------------------------------------

    @override
    async def run_async(self, *, args: dict[str, Any], tool_context: ToolContext) -> Any:
        """Execute the graph search with merged parameters."""
        # --- Resolve search target ------------------------------------
        search_kwargs: dict[str, Any] = {}

        if self._graph_id:
            search_kwargs["graph_id"] = self._graph_id
        else:
            user_id = self._resolve_user_id(tool_context)
            if not user_id:
                return "Error: Cannot determine user ID from session state."
            search_kwargs["user_id"] = user_id

        # --- Merge params: pinned > model-provided > default ----------
        for param_name, param_def in _SEARCH_PARAMS.items():
            if param_name in self._pinned:
                search_kwargs[param_name] = self._pinned[param_name]
            elif param_name in args:
                search_kwargs[param_name] = args[param_name]
            elif "default" in param_def:
                search_kwargs[param_name] = param_def["default"]
            # else: optional param not set by anyone → omit

        if not search_kwargs.get("query"):
            return "Error: No search query provided."

        # Constructor-only complex params
        for key in _CONSTRUCTOR_ONLY_PARAMS:
            if key in self._pinned:
                search_kwargs[key] = self._pinned[key]

        # --- Execute --------------------------------------------------
        scope = search_kwargs.get("scope", "edges")

        try:
            result = await self._zep.graph.search(**search_kwargs)
        except Exception as exc:
            logger.warning("Zep graph search failed: %s", exc, exc_info=True)
            return f"Graph search failed: {exc}"

        return self._format_results(result, scope)

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _resolve_user_id(tool_context: ToolContext) -> str | None:
        """Resolve user_id from session state or ADK session metadata."""
        state = tool_context.state
        if state is not None:
            user_id = state.get("zep_user_id")
            if user_id:
                return str(user_id)
        try:
            return tool_context._invocation_context.session.user_id  # type: ignore[union-attr]
        except AttributeError:
            return None

    @staticmethod
    def _format_results(result: Any, scope: str) -> str:
        """Format search results as readable text for the model."""
        parts: list[str] = []

        if scope == "edges" and result.edges:
            for edge in result.edges:
                if edge.fact:
                    parts.append(f"- {edge.fact}")
        elif scope == "nodes" and result.nodes:
            for node in result.nodes:
                name = getattr(node, "name", None) or "Entity"
                summary = getattr(node, "summary", None)
                if summary:
                    parts.append(f"- {name}: {summary}")
        elif scope == "episodes" and result.episodes:
            for ep in result.episodes:
                content = getattr(ep, "content", None)
                if content:
                    parts.append(f"- {content}")

        if not parts:
            return "No results found."

        return "\n".join(parts)

File: integrations/adk/typescript/src/graph-search-tool.ts
/**
 * `ZepGraphSearchTool` — a model-callable ADK tool for searching a Zep
 * knowledge graph on demand.
 *
 * Unlike `ZepContextTool` (which injects context automatically on every turn),
 * this tool appears in the model's tool list and is called when the model
 * decides it needs to look something up.
 *
 * Search target resolution:
 *   - If `graphUuid` is set, every search targets that graph.
 *   - Otherwise the graph of the current user is searched. The tool resolves
 *     the user UUID from explicit options → `zep_user_uuid` state key → the
 *     ADK `userId`, then reads the graph UUID of that user with
 *     `user.get(userUuid)` and caches it for the life of the tool.
 *
 * ## Pin-or-expose parameter model
 *
 * Every search knob (`scope`, `reranker`, `limit`, `mmrLambda`,
 * `centerNodeUuid`) is **tri-state** at construction time:
 *
 * - **Concrete value** (e.g. `scope: "edges"`) → *pinned*. Hidden from the
 *   model's tool schema and always used, regardless of what the model sends.
 * - **`null`** → *hidden*. Hidden from the model's tool schema AND omitted
 *   from the search call entirely (useful for suppressing optional params,
 *   e.g. `mmrLambda: null` when the reranker is never `"mmr"`).
 * - **`undefined` / omitted** → *exposed*. Included in the model's tool
 *   schema with the documented default, so the model can choose a value.
 *
 * `searchFilters` and `bfsOriginNodeUuids` are always constructor-only — they
 * are never exposed to the model and, when set, are always applied.
 */

import type { ZepClient, Zep } from "@getzep/zep-cloud";
import {
  BaseTool,
  type Context,
} from "@google/adk";
import { type FunctionDeclaration, type Schema, Type } from "@google/genai";
import {
  resolveIdentity,
  type AdkContextLike,
  type ZepIdentityOptions,
} from "./identity.js";
import { defaultLogger, type Logger } from "./logging.js";

const DEFAULT_DESCRIPTION =
  "Search the user's knowledge graph for information from previous " +
  "conversations, known facts about the user, or general context. " +
  "Use this to look up specific details the user has shared before.";

/**
 * The search scopes of the tool.
 *
 * Zep v4 has one search method for each scope, so the SDK no longer exports
 * a scope enum. The tool keeps the scope as its own parameter and calls the
 * matching v4 method.
 */
export type ZepGraphSearchScope =
  | "edges"
  | "nodes"
  | "episodes"
  | "observations"
  | "thread_summaries"
  | "auto";

/** The scopes that the tool can expose to the model. */
const SCOPE_ENUM = [
  "edges",
  "nodes",
  "episodes",
  "observations",
  "thread_summaries",
  "auto",
] as const;

/** Zep's supported rerankers (`V4SearchRequestReranker`). */
const RERANKER_ENUM = [
  "rrf",
  "mmr",
  "node_distance",
  "episode_mentions",
  "cross_encoder",
] as const;

/**
 * Search scopes this tool knows how to format, matching {@link SCOPE_ENUM}
 * exactly.
 */
const SUPPORTED_SCOPES = [
  "auto",
  "edges",
  "nodes",
  "episodes",
  "observations",
  "thread_summaries",
] as const satisfies readonly ZepGraphSearchScope[];

const DEFAULT_SCOPE: ZepGraphSearchScope = "edges";
const DEFAULT_RERANKER: Zep.V4SearchRequestReranker = "rrf";
const DEFAULT_LIMIT = 10;

/** Options for the {@link ZepGraphSearchTool} constructor. */
export interface ZepGraphSearchToolOptions extends ZepIdentityOptions {
  /** An initialised `ZepClient`. The caller owns its lifecycle. */
  zep: ZepClient;
  /**
   * Fixed graph UUID. When set, all searches target this graph regardless of
   * the active user. When omitted, the tool reads the graph UUID of the
   * resolved user from Zep and searches that graph.
   */
  graphUuid?: string;
  /** Tool name shown to the model. Defaults to `"zep_graph_search"`. */
  name?: string;
  /** Tool description shown to the model. */
  description?: string;
  /**
   * Search scope: `"edges"` for facts and relationships, `"nodes"` for
   * entities and their summaries, `"episodes"` for raw text data
   * (unstructured text, messages, or JSON), `"observations"` for derived
   * memories, `"thread_summaries"` for incremental thread summaries, `"auto"`
   * to let Zep decide the best mix of results.
   *
   * Tri-state: a concrete value pins the scope (hidden from the model,
   * always used); `null` hides it AND omits it from the search call;
   * `undefined`/omitted exposes it to the model, defaulting to `"edges"`.
   */
  scope?: ZepGraphSearchScope | null;
  /**
   * Result ordering algorithm: `"rrf"` (balanced), `"mmr"` (diverse),
   * `"cross_encoder"` (highest accuracy), `"episode_mentions"` (frequently
   * referenced), `"node_distance"` (near a specific entity).
   *
   * Tri-state: a concrete value pins the reranker (hidden from the model,
   * always used); `null` hides it AND omits it from the search call;
   * `undefined`/omitted exposes it to the model, defaulting to `"rrf"`.
   */
  reranker?: Zep.V4SearchRequestReranker | null;
  /**
   * Maximum number of results to return.
   *
   * Tri-state: a concrete value pins the limit (hidden from the model,
   * always used); `null` hides it AND omits it from the search call;
   * `undefined`/omitted exposes it to the model, defaulting to `10`.
   */
  limit?: number | null;
  /**
   * Balance between diversity (`0.0`) and relevance (`1.0`). Only used when
   * `reranker` is `"mmr"`.
   *
   * Tri-state: a concrete value pins `mmrLambda` (hidden from the model,
   * always used); `null` hides it AND omits it from the search call;
   * `undefined`/omitted exposes it to the model with no default.
   */
  mmrLambda?: number | null;
  /**
   * UUID of the center node for distance-based reranking. Required when
   * `reranker` is `"node_distance"`.
   *
   * Tri-state: a concrete value pins `centerNodeUuid` (hidden from the
   * model, always used); `null` hides it AND omits it from the search call;
   * `undefined`/omitted exposes it to the model with no default.
   */
  centerNodeUuid?: string | null;
  /**
   * Search filters to apply (`nodeLabels`, `edgeTypes`,
   * `excludeNodeLabels`, `excludeEdgeTypes`, property filters, etc.).
   * Constructor-only — never exposed to the model. Always applied when set.
   */
  searchFilters?: Zep.SearchFilters;
  /**
   * Node UUIDs seeding a breadth-first search. Constructor-only — never
   * exposed to the model. Always applied when set.
   */
  bfsOriginNodeUuids?: string[];
  /** Logger for Zep failures and invalid-argument warnings. Defaults to a `console`-backed logger. */
  logger?: Logger;
}

/** Arguments the model supplies when calling the tool. */
interface GraphSearchArgs {
  query?: unknown;
  scope?: unknown;
  reranker?: unknown;
  limit?: unknown;
  mmrLambda?: unknown;
  centerNodeUuid?: unknown;
}

/** Tri-state resolution for a single pinnable/exposable parameter. */
type PinState<T> =
  | { kind: "pinned"; value: T }
  | { kind: "hidden" }
  | { kind: "exposed" };

function resolvePinState<T>(value: T | null | undefined): PinState<T> {
  if (value === null) return { kind: "hidden" };
  if (value === undefined) return { kind: "exposed" };
  return { kind: "pinned", value };
}

/**
 * A model-callable tool that searches a Zep knowledge graph.
 *
 * Every search parameter (`scope`, `reranker`, `limit`, `mmrLambda`,
 * `centerNodeUuid`) can be pinned, hidden, or exposed to the model — see the
 * {@link ZepGraphSearchToolOptions} tri-state documentation. Only exposed
 * parameters appear in the model's tool schema; `query` is always required.
 *
 * Errors are returned to the model as text rather than thrown, so a Zep
 * failure cannot crash the agent run.
 *
 * @example
 * ```ts
 * import { LlmAgent } from "@google/adk";
 * import { ZepClient } from "@getzep/zep-cloud";
 * import { ZepGraphSearchTool } from "@getzep/zep-adk";
 *
 * const zep = new ZepClient({ apiKey: process.env.ZEP_API_KEY! });
 * const agent = new LlmAgent({
 *   name: "memory_agent",
 *   model: "gemini-2.5-flash",
 *   instruction: "You are a helpful assistant.",
 *   tools: [new ZepGraphSearchTool({ zep, userUuid })],
 * });
 * ```
 *
 * @example Restoring the old "always pinned" behavior
 * ```ts
 * // Model only ever sees `query`; scope/reranker/limit are fixed.
 * new ZepGraphSearchTool({
 *   zep,
 *   userUuid,
 *   scope: "edges",
 *   reranker: "rrf",
 *   limit: 10,
 *   mmrLambda: null,
 *   centerNodeUuid: null,
 * });
 * ```
 */
export class ZepGraphSearchTool extends BaseTool {
  private readonly zep: ZepClient;
  private readonly logger: Logger;
  private readonly graphUuid?: string;
  private readonly identity: ZepIdentityOptions;
  private readonly declaration: FunctionDeclaration;
  /** Graph UUID for each resolved user UUID, cached for the life of the tool. */
  private readonly graphUuidByUser = new Map<string, string>();

  private readonly scopeState: PinState<ZepGraphSearchScope>;
  private readonly rerankerState: PinState<Zep.V4SearchRequestReranker>;
  private readonly limitState: PinState<number>;
  private readonly mmrLambdaState: PinState<number>;
  private readonly centerNodeUuidState: PinState<string>;
  private readonly searchFilters?: Zep.SearchFilters;
  private readonly bfsOriginNodeUuids?: string[];

  constructor(options: ZepGraphSearchToolOptions) {
    super({
      name: options.name ?? "zep_graph_search",
      description: options.description ?? DEFAULT_DESCRIPTION,
    });
    this.zep = options.zep;
    this.logger = options.logger ?? defaultLogger;
    this.graphUuid = options.graphUuid;
    this.identity = {
      userUuid: options.userUuid,
      threadUuid: options.threadUuid,
      firstName: options.firstName,
      lastName: options.lastName,
    };

    this.scopeState = resolvePinState(options.scope);
    this.rerankerState = resolvePinState(options.reranker);
    this.limitState = resolvePinState(options.limit);
    this.mmrLambdaState = resolvePinState(options.mmrLambda);
    this.centerNodeUuidState = resolvePinState(options.centerNodeUuid);

    if (
      this.scopeState.kind === "pinned" &&
      !SUPPORTED_SCOPES.includes(this.scopeState.value)
    ) {
      throw new Error(
        `Unsupported Zep graph search scope: '${this.scopeState.value}'. ` +
          `Supported scopes are: ${SUPPORTED_SCOPES.join(", ")}.`,
      );
    }

    this.searchFilters = options.searchFilters;
    this.bfsOriginNodeUuids = options.bfsOriginNodeUuids;

    this.declaration = this.buildDeclaration();
  }

  /**
   * Builds the function declaration from exposed (non-pinned, non-hidden)
   * parameters only. `query` is always present and required.
   */
  private buildDeclaration(): FunctionDeclaration {
    const properties: Record<string, Schema> = {
      query: {
        type: Type.STRING,
        description: "Search query text (max 400 characters).",
      },
    };

    if (this.scopeState.kind === "exposed") {
      properties.scope = {
        type: Type.STRING,
        description:
          "What to search for: 'edges' for facts and relationships, " +
          "'nodes' for entities and their summaries, " +
          "'episodes' for raw text data (unstructured text, messages, or JSON), " +
          "'observations' for derived memories, " +
          "'thread_summaries' for incremental thread summaries, " +
          "'auto' to let Zep decide the best mix of results.",
        enum: [...SCOPE_ENUM],
      };
    }

    if (this.rerankerState.kind === "exposed") {
      properties.reranker = {
        type: Type.STRING,
        description:
          "Result ordering algorithm: 'rrf' (balanced), 'mmr' (diverse), " +
          "'cross_encoder' (highest accuracy), 'episode_mentions' " +
          "(frequently referenced), 'node_distance' (near a specific entity).",
        enum: [...RERANKER_ENUM],
      };
    }

    if (this.limitState.kind === "exposed") {
      properties.limit = {
        type: Type.INTEGER,
        description: "Maximum number of results to return.",
      };
    }

    if (this.mmrLambdaState.kind === "exposed") {
      properties.mmrLambda = {
        type: Type.NUMBER,
        description:
          "Balance between diversity (0.0) and relevance (1.0). Only used when reranker is 'mmr'.",
      };
    }

    if (this.centerNodeUuidState.kind === "exposed") {
      properties.centerNodeUuid = {
        type: Type.STRING,
        description:
          "UUID of the center node for distance-based reranking. " +
          "Required when reranker is 'node_distance'.",
      };
    }

    return {
      name: this.name,
      description: this.description,
      parameters: {
        type: Type.OBJECT,
        properties,
        required: ["query"],
      },
    };
  }

  /** Exposes the declaration built from non-pinned, non-hidden parameters. */
  override _getDeclaration(): FunctionDeclaration {
    return this.declaration;
  }

  /**
   * Run the graph search and return formatted results as text for the model.
   *
   * Returns a human-readable error string (never throws) when the search
   * target cannot be resolved or the Zep call fails.
   */
  override async runAsync(request: {
    args: Record<string, unknown>;
    toolContext: Context;
  }): Promise<unknown> {
    const { args, toolContext } = request;
    const typedArgs = args as GraphSearchArgs;
    const query = typedArgs.query;
    if (typeof query !== "string" || query.trim().length === 0) {
      return "Error: a non-empty 'query' string is required.";
    }

    const scope = this.resolveEnum(
      "scope",
      this.scopeState,
      typedArgs.scope,
      SCOPE_ENUM,
      DEFAULT_SCOPE,
    );
    const reranker = this.resolveEnum(
      "reranker",
      this.rerankerState,
      typedArgs.reranker,
      RERANKER_ENUM,
      DEFAULT_RERANKER,
    );
    const limit = this.resolveNumber(
      "limit",
      this.limitState,
      typedArgs.limit,
      DEFAULT_LIMIT,
    );
    const mmrLambda = this.resolveNumber(
      "mmrLambda",
      this.mmrLambdaState,
      typedArgs.mmrLambda,
    );
    const centerNodeUuid = this.resolveString(
      this.centerNodeUuidState,
      typedArgs.centerNodeUuid,
    );

    const body: Zep.SearchRequest = { query };
    if (reranker !== undefined) body.reranker = reranker;
    if (mmrLambda !== undefined) body.mmrLambda = mmrLambda;
    if (centerNodeUuid !== undefined) body.centerNodeUuid = centerNodeUuid;
    if (this.searchFilters !== undefined) {
      body.filters = this.searchFilters;
    }
    if (this.bfsOriginNodeUuids !== undefined) {
      body.bfsOriginNodeUuids = this.bfsOriginNodeUuids;
    }

    let graphUuid: string;
    try {
      graphUuid = await this.resolveGraphUuid(toolContext);
    } catch (error) {
      this.logger.warn("Could not resolve the graph to search", error);
      return "Error: could not determine which graph to search.";
    }

    const resolvedScope = scope ?? DEFAULT_SCOPE;
    try {
      const texts = await searchGraphScope(this.zep, graphUuid, {
        scope: resolvedScope,
        body,
        limit,
      });
      if (texts.length === 0) {
        return "No results found.";
      }
      // The "auto" scope returns one pre-assembled Context Block.
      return resolvedScope === "auto"
        ? texts[0]
        : texts.map((text) => `- ${text}`).join("\n");
    } catch (error) {
      this.logger.warn("Zep graph search failed", error);
      return `Graph search failed: ${error instanceof Error ? error.message : String(error)}`;
    }
  }

  /**
   * Resolve an enum-valued parameter: pinned beats model-provided arg beats
   * default. An invalid model-provided value falls back to the default with
   * a logged warning rather than throwing.
   */
  private resolveEnum<T extends string>(
    paramName: string,
    state: PinState<T>,
    modelValue: unknown,
    validValues: readonly T[],
    fallbackDefault: T,
  ): T | undefined {
    if (state.kind === "pinned") return state.value;
    if (state.kind === "hidden") return undefined;
    if (typeof modelValue === "string") {
      if ((validValues as readonly string[]).includes(modelValue)) {
        return modelValue as T;
      }
      this.logger.warn(
        `Zep graph search: invalid '${paramName}' value from model: ${JSON.stringify(
          modelValue,
        )}. Falling back to default '${fallbackDefault}'.`,
      );
      return fallbackDefault;
    }
    return fallbackDefault;
  }

  /**
   * Resolve a numeric parameter: pinned beats model-provided arg beats
   * default. A non-numeric model-provided value falls back to the default
   * (or is omitted, if there is none) with a logged warning.
   */
  private resolveNumber(
    paramName: string,
    state: PinState<number>,
    modelValue: unknown,
    fallbackDefault?: number,
  ): number | undefined {
    if (state.kind === "pinned") return state.value;
    if (state.kind === "hidden") return undefined;
    if (modelValue !== undefined) {
      if (typeof modelValue === "number" && Number.isFinite(modelValue)) {
        return modelValue;
      }
      this.logger.warn(
        `Zep graph search: invalid '${paramName}' value from model: ${JSON.stringify(
          modelValue,
        )}. Falling back to default.`,
      );
      return fallbackDefault;
    }
    return fallbackDefault;
  }

  /**
   * Resolve a string parameter with no default: pinned beats model-provided
   * arg beats "unset" (`undefined`, meaning omitted from the SDK call).
   */
  private resolveString(
    state: PinState<string>,
    modelValue: unknown,
  ): string | undefined {
    if (state.kind === "pinned") return state.value;
    if (state.kind === "hidden") return undefined;
    if (typeof modelValue === "string" && modelValue.length > 0) {
      return modelValue;
    }
    return undefined;
  }

  /**
   * Resolve the UUID of the graph to search.
   *
   * A configured `graphUuid` wins. Otherwise the tool resolves the user UUID
   * of the turn and reads the graph UUID of that user once, then serves
   * later turns of the same user from its cache. The tool never calls
   * `user.lookup`: a `userId` name is not an address in v4.
   */
  private async resolveGraphUuid(context: AdkContextLike): Promise<string> {
    if (this.graphUuid) {
      return this.graphUuid;
    }

    const { userUuid } = resolveIdentity(context, this.identity);
    const cached = this.graphUuidByUser.get(userUuid);
    if (cached) {
      return cached;
    }

    const user = await this.zep.user.get(userUuid);
    if (!user.graphUuid) {
      throw new Error(`Zep user ${userUuid} has no graph UUID.`);
    }
    this.graphUuidByUser.set(userUuid, user.graphUuid);
    return user.graphUuid;
  }
}

/**
 * Joins a name and summary as `"name: summary"`, falling back to whichever
 * half is present, or `""` when both are empty/undefined. Mirrors Go's
 * `nameSummaryText` (see `integrations/adk/go/search.go`), which Python's
 * `_format_results` also matches for the observations/thread_summaries cases.
 */
function nameSummaryText(
  name: string | null | undefined,
  summary: string | null | undefined,
): string {
  if (name && summary) return `${name}: ${summary}`;
  if (name) return name;
  return summary ?? "";
}

/** Parameters of a single scoped graph search. */
export interface GraphScopeSearchParams {
  /** The scope to search. */
  scope: ZepGraphSearchScope;
  /** The v4 search body. The `"auto"` scope uses its `query` and `filters`. */
  body: Zep.SearchRequest;
  /**
   * Maximum number of results on the first page. Zep v4 has no result limit
   * for the `"auto"` scope, so the value is ignored for that scope.
   */
  limit?: number;
}

/**
 * Runs one scoped v4 graph search and returns plain-text items.
 *
 * Shared by {@link ZepGraphSearchTool} (which prefixes each item with `"- "`
 * for the model) and `ZepMemoryService` (which wraps each item in its own
 * `MemoryEntry`).
 *
 * Zep v4 replaces the single v3 `graph.search` method with one method for
 * each scope, and the `"auto"` scope with `graph.getContext`. The v4 search
 * methods return one page of results with an opaque cursor; this function
 * reads the first page only, as the v3 call did.
 */
export async function searchGraphScope(
  zep: ZepClient,
  graphUuid: string,
  params: GraphScopeSearchParams,
): Promise<string[]> {
  const { scope, body, limit } = params;

  if (scope === "auto") {
    const response = await zep.graph.getContext(graphUuid, {
      query: body.query,
      filters: body.filters,
    });
    const context = response.context?.trim();
    return context ? [context] : [];
  }

  const request = { limit, body };
  const texts: string[] = [];

  if (scope === "edges") {
    const page = await zep.graph.searchEdges(graphUuid, request);
    for (const edge of page.data) {
      if (edge.fact) texts.push(edge.fact);
    }
  } else if (scope === "nodes") {
    const page = await zep.graph.searchNodes(graphUuid, request);
    for (const node of page.data) {
      const text = nameSummaryText(node.name, node.summary);
      if (text) texts.push(text);
    }
  } else if (scope === "episodes") {
    const page = await zep.graph.searchEpisodes(graphUuid, request);
    for (const episode of page.data) {
      if (episode.content) texts.push(episode.content);
    }
  } else if (scope === "observations") {
    const page = await zep.graph.searchObservations(graphUuid, request);
    for (const observation of page.data) {
      const text = nameSummaryText(observation.name, observation.summary);
      if (text) texts.push(text);
    }
  } else {
    const page = await zep.graph.searchThreadSummaries(graphUuid, request);
    for (const threadSummary of page.data) {
      const text = threadSummary.summary?.trim();
      if (text) texts.push(text);
    }
  }

  return texts;
}
